### PR TITLE
refactor snapshot controller

### PR DESCRIPTION
**Pull request by mhenriks**
Signed-off-by: Michael Henriksen <mhenriks@redhat.com>

**What this PR does / why we need it**:

Refactor to hopefully make code easier to follow by reducing paths/calls to Update.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 1]
```diff
@@ -230,11 +212,25 @@ func (ctrl *VMSnapshotController) updateVMSnapshotContent(content *snapshotv1.Vi
...
+			if err != nil {
+				return 0, err
+			}
+
+			frozen, err := source.Frozen()
 			if err != nil {
 				return 0, err
 			}
 
+			if !frozen {
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:225]: freeze should only be when its online vm snapshot and you didnt check here if the vm is online

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 2]
```diff
@@ -279,6 +275,24 @@ func (ctrl *VMSnapshotController) updateVMSnapshotContent(content *snapshotv1.Vi
 
 	if ready && (contentCpy.Status.ReadyToUse == nil || !*contentCpy.Status.ReadyToUse) {
 		contentCpy.Status.CreationTime = currentTime()
+
+		source, err := ctrl.getSnapshotSource(vmSnapshot)
+		if err != nil {
+			return 0, err
+		}
+
+		frozen, err := source.Frozen()
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:284]: This code will only happen once, if the vm status havnt changed yet to frozen (which can take up to 5 secs) then this will be false and the unfrozen will never occure

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 3]
```diff
@@ -279,6 +275,24 @@ func (ctrl *VMSnapshotController) updateVMSnapshotContent(content *snapshotv1.Vi
...
+		if err != nil {
+			return 0, err
+		}
+
+		if frozen {
+			if err := source.Unfreeze(); err != nil {
+				return 0, err
+			}
+
+			ready = false
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:294]: why change ready to false?

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 4]
```diff
@@ -647,6 +634,41 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
...
+					return err
+				}
+
+				if frozen {
+					updateSnapshotCondition(vmSnapshotCpy, newFreezingCondition(corev1.ConditionTrue, vmSnapshotSourceFrozen))
+				} else {
+					updateSnapshotCondition(vmSnapshotCpy, newFreezingCondition(corev1.ConditionFalse, vmSnapshotSourceThawed))
+				}
+			} else {
+				updateSnapshotCondition(vmSnapshotCpy, newFreezingCondition(corev1.ConditionFalse, vmSnapshotSourceThawed))
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:670]: why adding the freezing condition if the snapshot is not online? although it is true that the source is thawed since we never freezed but seems irrelevant in such case

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 5]
```diff
@@ -657,6 +679,7 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
 	} else if vmSnapshotReady(vmSnapshotCpy) {
 		updateSnapshotCondition(vmSnapshotCpy, newProgressingCondition(corev1.ConditionFalse, "Operation complete"))
 		updateSnapshotCondition(vmSnapshotCpy, newReadyCondition(corev1.ConditionTrue, "Operation complete"))
+		updateSnapshotCondition(vmSnapshot, newFreezingCondition(corev1.ConditionFalse, vmSnapshotSourceThawed))
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:682]: same here, also if the vm is still frozen because of some bug that the unfreeze never happened then this is actually wrong

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 6]
```diff
@@ -647,6 +634,41 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
...
+
+				if ga {
+					indications = append(indications, snapshotv1.VMSnapshotGuestAgentIndication)
+				} else {
+					indications = append(indications, snapshotv1.VMSnapshotNoGuestAgentIndication)
+				}
+
+				vmSnapshotCpy.Status.Indications = indications
+
+				frozen, err := source.Frozen()
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:659]: using this state as an indication for the source state is not so accurate since the status is updated only every 5 secs so it can be already frozen and this will return false, so the frozen condition might never be updated to true. I added those conditions to use to know if unfreeze is needed and if happened it is maybe redundant with this repactor

File: pkg/virt-controller/watch/snapshot/source.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+	return condManager.HasCondition(vmi, kubevirtv1.VirtualMachineInstanceAgentConnected), nil
+}
+
+func (s *vmSnapshotSource) Frozen() (bool, error) {
+	vmi, exists, err := s.controller.getVMI(s.vm)
+	if err != nil || !exists {
+		return false, err
+	}
+
+	return vmi.Status.FSFreezeStatus == "frozen", nil
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/source.go:187]: I added a const variable for "frozen", not sure if in the merged PR or the other PR its in api.FSFROZEN

File: pkg/virt-controller/watch/snapshot/source.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+func (s *vmSnapshotSource) Freeze() error {
+	exists, err := s.GuestAgent()
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/source.go:191]: I commented on the fact that the freeze should only be on online vm in other comment. maybe here you can add the check of s.Online()

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 2 above)
Comment (ShellyKa13) [reply pkg/virt-controller/watch/snapshot/snapshot.go:284]: maybe if online and guest agent you can assume its frozen at this point and just unfreeze it wont fail if it is already unfrozen..

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 1 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:225]: as-is, freeze will return nil if offline (no VMI) but is probably better to be explicit

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 2 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:284]: Ah right since we are not waiting to observe the VMI being frozen we may miss the unfreeze here hmmmm

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 3 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:294]: THe idea is to actually observe the VM being thawed before saying ready to use

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 6 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:659]: Understood.  Not sure it's a big deal though

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 4 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:670]: yeah could go either way on this.  the ocd part of me wants to include everything but maybe it's confusing?

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 5 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:682]: The idea is that the content will not be ready (hence the snapshot) until observed unfrozen

File: pkg/virt-controller/watch/snapshot/source.go (hunk 8 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/source.go:191]: THat would be more through.  but guestagent will return false if no VMI

Comment (ShellyKa13) [reply pkg/virt-controller/watch/snapshot/source.go:191]: right! so ignore my comments regarding the online :)

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 5 above)
Comment (ShellyKa13) [reply pkg/virt-controller/watch/snapshot/snapshot.go:682]: thats the ideal haha but it might not be so - i.e bug

**Review by davidvossel — COMMENTED** [review]
way better. The only thing that kind of throws me for a loop is the finalizer logic.

We know that updateVMSnapshot() will get called before the snapshot is ever worked on, and we know that the finalizer can be removed during deletion when content == nil.


so, in updateVMSnapshot, why not unconditionally add the finalizer when vmSnapshot.DeletionTimestamp == nil and unconditionally remove it when `vmSnapshot.DeletionTimestamp != nil && content == nil || deletion is set to retain and content finalizer is removed`? Then all the logic around the finalizer centralized.

File: tests/storage/restore.go [hunk 9]
```diff
@@ -229,6 +229,10 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
 					vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
 					Expect(err).ToNot(HaveOccurred())
 
+					if vm.Status.SnapshotInProgress != nil {
```
Comment (ShellyKa13) [inline tests/storage/restore.go:232]: I think the PR I did the was merged yesterday fixes this: https://github.com/kubevirt/kubevirt/pull/6146

File: tests/storage/snapshot.go [hunk 10]
```diff
@@ -312,10 +301,13 @@ var _ = SIGDescribe("[Serial]VirtualMachineSnapshot Tests", func() {
 						&expect.BExp{R: console.PromptExpression},
 						&expect.BSnd{S: "echo $?\n"},
 						&expect.BExp{R: console.RetValue("1")},
-						&expect.BSnd{S: fmt.Sprintf("%s | grep \"%s\"\n", journalctlCheck, expectedThawOutput)},
-						&expect.BExp{R: console.PromptExpression},
-						&expect.BSnd{S: "echo $?\n"},
-						&expect.BExp{R: console.RetValue("1")},
+						// always thaw
```
Comment (ShellyKa13) [inline tests/storage/snapshot.go:304]: it might call thaw :) since we do the freeze only if we have volume backups but we thaw if we have guest agent

File: tests/storage/snapshot.go [hunk 11]
```diff
@@ -257,37 +256,27 @@ var _ = SIGDescribe("[Serial]VirtualMachineSnapshot Tests", func() {
...
-					conditionsLength = 2
-				}
+				conditionsLength := 2
 				Expect(snapshot.Status.Conditions).To(HaveLen(conditionsLength))
 				Expect(snapshot.Status.Conditions[0].Type).To(Equal(snapshotv1.ConditionProgressing))
 				Expect(snapshot.Status.Conditions[0].Status).To(Equal(corev1.ConditionFalse))
 				Expect(snapshot.Status.Conditions[1].Type).To(Equal(snapshotv1.ConditionReady))
 				Expect(snapshot.Status.Conditions[1].Status).To(Equal(corev1.ConditionTrue))
+
 				if shouldFreeze {
```
Comment (ShellyKa13) [inline tests/storage/snapshot.go:275]: this check is now true either way the fsfreezestatus after the snapshot should always be ""

File: tests/storage/snapshot.go [hunk 12]
```diff
@@ -516,21 +508,16 @@ var _ = SIGDescribe("[Serial]VirtualMachineSnapshot Tests", func() {
 				tests.WaitAgentConnected(virtClient, vmi)
 
 				createDenyVolumeSnapshotCreateWebhook()
+				defer deleteWebhook()
```
Comment (ShellyKa13) [inline tests/storage/snapshot.go:511]: isn't it the redundant as it happens in the AfterTest?

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 13]
```diff
@@ -607,6 +605,13 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
 	if vmSnapshotCpy.DeletionTimestamp != nil {
 		// go into error state
 		if vmSnapshotProgressing(vmSnapshotCpy) {
+			if source != nil {
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/snapshot.go:608]: I think it is better to do this code in the main updatesnaphot before the vmSnapshotCpy.DeletionTimestamp != nil, since it will cover also the deadline, in both cases we mark error on the snapshot which makes vmSnapshotProgressing(vmSnapshotCpy) to be false and enter the if in the reconcile loop

File: tests/storage/snapshot.go (hunk 12 above)
Comment (mhenriks) [reply tests/storage/snapshot.go:511]: wasn't happening in my testing

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 13 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:608]: was thinking we may actually want to handle deadline in the snapshotcontent since that is where we freeze in the first place.  what do you think?

File: tests/storage/snapshot.go (hunk 11 above)
Comment (mhenriks) [reply tests/storage/snapshot.go:275]: good call

File: tests/storage/snapshot.go (hunk 10 above)
Comment (mhenriks) [reply tests/storage/snapshot.go:304]: maybe use new hasGuestAgent flag here then?

File: tests/storage/restore.go (hunk 9 above)
Comment (mhenriks) [reply tests/storage/restore.go:232]: ah right don't think this hurts anything though

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 13 above)
Comment (ShellyKa13) [reply pkg/virt-controller/watch/snapshot/snapshot.go:608]: I dont think so, it is relevant for every snapshot not only with freeze.. also it is relevant for the snapshot as a whole, we compare the timeout to the time the snapshot was created

Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:608]: Good call

**Comment by mhenriks** [comment]
@davidvossel 
>We know that updateVMSnapshot() will get called before the snapshot is ever worked on, and we know that the finalizer can be removed during deletion when content == nil.

>so, in updateVMSnapshot, why not unconditionally add the finalizer when vmSnapshot.DeletionTimestamp == nil and unconditionally remove it when vmSnapshot.DeletionTimestamp != nil && content == nil || deletion is set to retain and content finalizer is removed? Then all the logic around the finalizer centralized.

Maybe it does make sense to to add the finalizer to the vmsnapshot it sooner/more obvious place

But have to be kind of careful with removing the finalizer.  Don't want to remove it when the source is locked/frozen.  So what happens now is that `updateSnapshotStatus` checks the deletion timestamp and will set an error on the snapshot.  Then we reconcile again and `updateVMSnapshot` sees in error state and does all cleanup in one place including removing finalizers but only after sources are thawed/unlocked

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve
